### PR TITLE
Fix null ptr deref when renaming a zvol with snaps and snapdev=visible

### LINKED ISSUE #16274: Kernel oops (null pointer dereference) when renaming a zvol
### System information
Type | Version/Name
 --- | ---
Distribution Name       | Arch Linux
Distribution Version    | Rolling release (packages updated Jun 1 2024)
Kernel Version  | 6.9.2-arch1-1
Architecture    | x86_64
OpenZFS Version | 2.2.4 (stable)

### Describe the problem you're observing
Doing `zfs rename` on a zvol causes a kernel oops, which kills kthread `z_zvol` and makes zvols unusable until the system is rebooted.

The kernel oops is a null pointer dereference to `0x120` in `dataset_kstats_rename`. This happens on the second of the two CPU instructions for `dk->dk_kstats->ks_data`, in which `ks_data` is accessed by dereferencing `dk_kstats`. `dk->dk_kstats` is null; and `offsetof(kstat_t, ks_data) == 0x120` (in this particular build, anyway); and so therefore the taskq thread tries to access the qword at address `0x120`.

```c
void
dataset_kstats_rename(dataset_kstats_t *dk, const char *name)
{
        dataset_kstat_values_t *dkv = dk->dk_kstats->ks_data; // <-- dk->dk_kstats is null
        char *ds_name;

        ds_name = KSTAT_NAMED_STR_PTR(&dkv->dkv_ds_name);
        ASSERT3S(ds_name, !=, NULL);
        (void) strlcpy(ds_name, name,
            KSTAT_NAMED_STR_BUFLEN(&dkv->dkv_ds_name));
}
```

```
00000000000eb840 <dataset_kstats_rename>:
   eb840:       f3 0f 1e fa             endbr64
   eb844:       e8 00 00 00 00          call   eb849 <dataset_kstats_rename+0x9>        eb845: R_X86_64_PLT32   __fentry__-0x4
   eb849:       48 8b 87 98 03 00 00    mov    rax,QWORD PTR [rdi+0x398]  <-- rax (aka: dk->dk_kstats) is zero after this instruction
   eb850:       48 8b 80 20 01 00 00    mov    rax,QWORD PTR [rax+0x120]  <-- this attempts to read qword ptr [0+0x120] and fails
   eb857:       8b 90 08 01 00 00       mov    edx,DWORD PTR [rax+0x108]
   eb85d:       48 8b b8 00 01 00 00    mov    rdi,QWORD PTR [rax+0x100]
   eb864:       e9 00 00 00 00          jmp    eb869 <dataset_kstats_rename+0x29>       eb865: R_X86_64_PLT32   sized_strscpy-0x4
```

### Describe how to reproduce the problem
1. Use OpenZFS 2.2.4 on Linux
2. Create a pool
3. Create a zvol
4. Do `zfs rename pool/path/to/zvol pool/path/to/newname`
5. Check `dmesg` and see if things went wrong

I updated OpenZFS on this system to 2.2.4, after being on an older version for quite some time, and I did that in preparation for creating an entirely new pool, which I've been using pretty extensively the past couple of weeks. During that time, I've run into this particular issue twice, and I think it has actually happened every time I have specifically done `zfs rename` on a zvol. (And actually yeah: I just checked `zpool history` for the relevant pool, and the two kernel oopses happened simultaneous with the only two times when I have done `zfs rename` on a zvol.)

I'm not yet sure whether maybe there are more conditions to the problem manifesting than literally just "use 2.2.4 and do `zfs rename` on a zvol". There was a reasonable amount of other, non-zvol, I/O happening with the pool at both times when the problem occurred.

I haven't had an opportunity yet to do a bunch of test-reboot-test-reboot cycles on this; and rebooting is necessary because, as soon as the problem occurs, the `z_zvol` kernel thread is murdered (while holding one or more locks, I believe, or else leaving some other thread of execution forever waiting on the task completion); and from that point on, although most other ZFS things do continue to work, basically anything zvol-related will simply get stuck and make no forward progress.

### Include any warning/errors/backtraces from the system logs
<details>
<summary>Kernel Oops 1</summary>

```
2024-06-11T23:42:54-07:00 helium kernel: BUG: kernel NULL pointer dereference, address: 0000000000000120
2024-06-11T23:42:54-07:00 helium kernel: #PF: supervisor read access in kernel mode
2024-06-11T23:42:54-07:00 helium kernel: #PF: error_code(0x0000) - not-present page
2024-06-11T23:42:54-07:00 helium kernel: PGD 0 P4D 0 
2024-06-11T23:42:54-07:00 helium kernel: Oops: 0000 [#1] PREEMPT SMP PTI
2024-06-11T23:42:54-07:00 helium kernel: CPU: 0 PID: 974 Comm: z_zvol Tainted: P           OE      6.9.2-arch1-1 #1 0c9f583c4dce67c6fc47fa44804d20ac179c7f14
2024-06-11T23:42:54-07:00 helium kernel: Hardware name: Gigabyte Technology Co., Ltd. EP35-DS3R/EP35-DS3R, BIOS F4 06/19/2009
2024-06-11T23:42:54-07:00 helium kernel: RIP: 0010:dataset_kstats_rename+0x10/0x30 [zfs]
2024-06-11T23:42:54-07:00 helium kernel: Code: 1f 84 00 00 00 00 00 0f 1f 00 90 90 90 90 90 90 90 90 90 90 90 90 90 90 90 90 66 0f 1f 00 0f 1f 44 00 00 48 8b 87 98 03 00 00 <48> 8b 80 20 01 00 00 8b 90 08 01 00 00 48 8b b8 00 01 00 00 e9 67
2024-06-11T23:42:54-07:00 helium kernel: RSP: 0018:ffffb1ee83fcfd38 EFLAGS: 00010292
2024-06-11T23:42:54-07:00 helium kernel: RAX: 0000000000000000 RBX: ffff99ea9ce1c000 RCX: 00000000ab9b9c00
2024-06-11T23:42:54-07:00 helium kernel: RDX: 00000000ab9b9a00 RSI: ffff99ea8d93e000 RDI: ffff99ea89167998
2024-06-11T23:42:54-07:00 helium kernel: RBP: ffff99ea89161800 R08: 0000000000000000 R09: 0000000000000000
2024-06-11T23:42:54-07:00 helium kernel: R10: 0000000000000000 R11: 0000000000000000 R12: ffff99ea89167d60
2024-06-11T23:42:54-07:00 helium kernel: R13: ffff99ea9ce1c004 R14: ffff99ea55ae3180 R15: ffff99ea89167800
2024-06-11T23:42:54-07:00 helium kernel: FS:  0000000000000000(0000) GS:ffff99eb41c00000(0000) knlGS:0000000000000000
2024-06-11T23:42:54-07:00 helium kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
2024-06-11T23:42:54-07:00 helium kernel: CR2: 0000000000000120 CR3: 0000000003b88000 CR4: 00000000000406f0
2024-06-11T23:42:54-07:00 helium kernel: Call Trace:
2024-06-11T23:42:54-07:00 helium kernel:  <TASK>
2024-06-11T23:42:54-07:00 helium kernel:  ? __die_body.cold+0x19/0x27
2024-06-11T23:42:54-07:00 helium kernel:  ? page_fault_oops+0x15a/0x2b0
2024-06-11T23:42:54-07:00 helium kernel:  ? search_bpf_extables+0x5f/0x80
2024-06-11T23:42:54-07:00 helium kernel:  ? exc_page_fault+0x81/0x190
2024-06-11T23:42:54-07:00 helium kernel:  ? asm_exc_page_fault+0x26/0x30
2024-06-11T23:42:54-07:00 helium kernel:  ? dataset_kstats_rename+0x10/0x30 [zfs 2d360922e0aeba07e78067b282e3eedb08db1dc5]
2024-06-11T23:42:54-07:00 helium kernel:  zvol_task_cb+0x37f/0x450 [zfs 2d360922e0aeba07e78067b282e3eedb08db1dc5]
2024-06-11T23:42:54-07:00 helium kernel:  taskq_thread+0x2e1/0x550 [spl 8d5c82e0079bc7d3bed9092345840f4851690968]
2024-06-11T23:42:54-07:00 helium kernel:  ? __pfx_default_wake_function+0x10/0x10
2024-06-11T23:42:54-07:00 helium kernel:  ? __pfx_taskq_thread+0x10/0x10 [spl 8d5c82e0079bc7d3bed9092345840f4851690968]
2024-06-11T23:42:54-07:00 helium kernel:  kthread+0xd2/0x100
2024-06-11T23:42:54-07:00 helium kernel:  ? __pfx_kthread+0x10/0x10
2024-06-11T23:42:54-07:00 helium kernel:  ret_from_fork+0x34/0x50
2024-06-11T23:42:54-07:00 helium kernel:  ? __pfx_kthread+0x10/0x10
2024-06-11T23:42:54-07:00 helium kernel:  ret_from_fork_asm+0x1a/0x30
2024-06-11T23:42:54-07:00 helium kernel:  </TASK>
2024-06-11T23:42:54-07:00 helium kernel: Modules linked in: tls rpcrdma rdma_cm iw_cm ib_cm ib_core cls_u32 sch_htb ifb wireguard curve25519_x86_64 libchacha20poly1305 chacha_x86_64 poly1305_x86_64 libcurve25519_generic libchacha ip6_udp_tunnel udp_tunnel bridge stp llc cfg80211 rfkill ip6table_nat ip6table_mangle ip6t_REJECT nf_reject_ipv6 xt_nat xt_REDIRECT xt_MASQUERADE iptable_nat xt_connmark xt_TCPMSS iptable_mangle xt_multiport xt_helper ipt_REJECT nf_reject_ipv4 xt_NFLOG nfnetlink_log xt_limit xt_conntrack xt_tcpudp nf_nat_irc nf_conntrack_irc nf_nat_sip nf_conntrack_sip nf_nat_ftp nf_conntrack_ftp nf_nat_pptp nf_conntrack_pptp nf_nat nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 lz4hc lz4hc_compress ip6table_filter ip6_tables iptable_filter coretemp kvm_intel iTCO_wdt intel_pmc_bxt ppdev gpio_ich iTCO_vendor_support zram kvm i2c_i801 sha512_ssse3 acpi_cpufreq sha1_ssse3 i2c_smbus lpc_ich intel_agp parport_pc intel_gtt parport joydev mac_hid zfs(POE) spl(OE) vhost_scsi target_core_mod vhost_net tun vhost vhost_iotlb tap usbip_host usbip_core
2024-06-11T23:42:54-07:00 helium kernel:  pcspkr it87(OE) nfsd hwmon_vid frandom(OE) sg auth_rpcgss crypto_user loop nfs_acl lockd grace sunrpc nfnetlink ip_tables x_tables hid_logitech_hidpp vfat fat hid_logitech_dj btrfs blake2b_generic ata_generic pata_acpi hid_generic xor raid6_pq usbhid sha256_ssse3 ext4 pata_jmicron crc16 mbcache jbd2 dm_cache dm_thin_pool dm_persistent_data dm_bio_prison libcrc32c crc32c_generic dm_snapshot dm_bufio dm_mod radeon drm_ttm_helper ttm video wmi i2c_algo_bit drm_suballoc_helper drm_display_helper cec lz4 lz4_compress netconsole r8169 realtek mdio_devres libphy r8168(OE)
2024-06-11T23:42:54-07:00 helium kernel: CR2: 0000000000000120
2024-06-11T23:42:54-07:00 helium kernel: ---[ end trace 0000000000000000 ]---
2024-06-11T23:42:54-07:00 helium kernel: RIP: 0010:dataset_kstats_rename+0x10/0x30 [zfs]
2024-06-11T23:42:54-07:00 helium kernel: Code: 1f 84 00 00 00 00 00 0f 1f 00 90 90 90 90 90 90 90 90 90 90 90 90 90 90 90 90 66 0f 1f 00 0f 1f 44 00 00 48 8b 87 98 03 00 00 <48> 8b 80 20 01 00 00 8b 90 08 01 00 00 48 8b b8 00 01 00 00 e9 67
2024-06-11T23:42:54-07:00 helium kernel: RSP: 0018:ffffb1ee83fcfd38 EFLAGS: 00010292
2024-06-11T23:42:54-07:00 helium kernel: RAX: 0000000000000000 RBX: ffff99ea9ce1c000 RCX: 00000000ab9b9c00
2024-06-11T23:42:54-07:00 helium kernel: RDX: 00000000ab9b9a00 RSI: ffff99ea8d93e000 RDI: ffff99ea89167998
2024-06-11T23:42:54-07:00 helium kernel: RBP: ffff99ea89161800 R08: 0000000000000000 R09: 0000000000000000
2024-06-11T23:42:54-07:00 helium kernel: R10: 0000000000000000 R11: 0000000000000000 R12: ffff99ea89167d60
2024-06-11T23:42:54-07:00 helium kernel: R13: ffff99ea9ce1c004 R14: ffff99ea55ae3180 R15: ffff99ea89167800
2024-06-11T23:42:54-07:00 helium kernel: FS:  0000000000000000(0000) GS:ffff99eb41c00000(0000) knlGS:0000000000000000
2024-06-11T23:42:54-07:00 helium kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
2024-06-11T23:42:54-07:00 helium kernel: CR2: 0000000000000120 CR3: 0000000003b88000 CR4: 00000000000406f0
2024-06-11T23:42:54-07:00 helium kernel: note: z_zvol[974] exited with irqs disabled
```
</details>
<details>
<summary>Kernel Oops 2</summary>

```
2024-06-16T16:09:44-07:00 helium kernel: BUG: kernel NULL pointer dereference, address: 0000000000000120
2024-06-16T16:09:44-07:00 helium kernel: #PF: supervisor read access in kernel mode
2024-06-16T16:09:44-07:00 helium kernel: #PF: error_code(0x0000) - not-present page
2024-06-16T16:09:44-07:00 helium kernel: PGD 0 P4D 0 
2024-06-16T16:09:44-07:00 helium kernel: Oops: 0000 [#1] PREEMPT SMP PTI
2024-06-16T16:09:44-07:00 helium kernel: CPU: 1 PID: 982 Comm: z_zvol Tainted: P           OE      6.9.2-arch1-1 #1 0c9f583c4dce67c6fc47fa44804d20ac179c7f14
2024-06-16T16:09:44-07:00 helium kernel: Hardware name: Gigabyte Technology Co., Ltd. EP35-DS3R/EP35-DS3R, BIOS F4 06/19/2009
2024-06-16T16:09:44-07:00 helium kernel: RIP: 0010:dataset_kstats_rename+0x10/0x30 [zfs]
2024-06-16T16:09:44-07:00 helium kernel: Code: 1f 84 00 00 00 00 00 0f 1f 00 90 90 90 90 90 90 90 90 90 90 90 90 90 90 90 90 66 0f 1f 00 0f 1f 44 00 00 48 8b 87 98 03 00 00 <48> 8b 80 20 01 00 00 8b 90 08 01 00 00 48 8b b8 00 01 00 00 e9 67
2024-06-16T16:09:44-07:00 helium kernel: RSP: 0018:ffffa06e886a3d38 EFLAGS: 00010292
2024-06-16T16:09:44-07:00 helium kernel: RAX: 0000000000000000 RBX: ffff88edacc7f000 RCX: 00000001861b8801
2024-06-16T16:09:44-07:00 helium kernel: RDX: 00000001861b8601 RSI: ffff88ee4072b100 RDI: ffff88ee212a7998
2024-06-16T16:09:44-07:00 helium kernel: RBP: ffff88edefc05000 R08: 0000000000000000 R09: 0000000000000000
2024-06-16T16:09:44-07:00 helium kernel: R10: 0000000000000000 R11: 0000000000000000 R12: ffff88ee212a7d60
2024-06-16T16:09:44-07:00 helium kernel: R13: ffff88edacc7f004 R14: ffff88ed96915280 R15: ffff88ee212a7800
2024-06-16T16:09:44-07:00 helium kernel: FS:  0000000000000000(0000) GS:ffff88ee81c80000(0000) knlGS:0000000000000000
2024-06-16T16:09:44-07:00 helium kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
2024-06-16T16:09:44-07:00 helium kernel: CR2: 0000000000000120 CR3: 000000004f110000 CR4: 00000000000406f0
2024-06-16T16:09:44-07:00 helium kernel: Call Trace:
2024-06-16T16:09:44-07:00 helium kernel:  <TASK>
2024-06-16T16:09:44-07:00 helium kernel:  ? __die_body.cold+0x19/0x27
2024-06-16T16:09:44-07:00 helium kernel:  ? page_fault_oops+0x15a/0x2b0
2024-06-16T16:09:44-07:00 helium kernel:  ? search_bpf_extables+0x5f/0x80
2024-06-16T16:09:44-07:00 helium kernel:  ? exc_page_fault+0x81/0x190
2024-06-16T16:09:44-07:00 helium kernel:  ? asm_exc_page_fault+0x26/0x30
2024-06-16T16:09:44-07:00 helium kernel:  ? dataset_kstats_rename+0x10/0x30 [zfs 2d360922e0aeba07e78067b282e3eedb08db1dc5]
2024-06-16T16:09:44-07:00 helium kernel:  zvol_task_cb+0x37f/0x450 [zfs 2d360922e0aeba07e78067b282e3eedb08db1dc5]
2024-06-16T16:09:44-07:00 helium kernel:  taskq_thread+0x2e1/0x550 [spl 8d5c82e0079bc7d3bed9092345840f4851690968]
2024-06-16T16:09:44-07:00 helium kernel:  ? __pfx_default_wake_function+0x10/0x10
2024-06-16T16:09:44-07:00 helium kernel:  ? __pfx_taskq_thread+0x10/0x10 [spl 8d5c82e0079bc7d3bed9092345840f4851690968]
2024-06-16T16:09:44-07:00 helium kernel:  kthread+0xd2/0x100
2024-06-16T16:09:44-07:00 helium kernel:  ? __pfx_kthread+0x10/0x10
2024-06-16T16:09:44-07:00 helium kernel:  ret_from_fork+0x34/0x50
2024-06-16T16:09:44-07:00 helium kernel:  ? __pfx_kthread+0x10/0x10
2024-06-16T16:09:44-07:00 helium kernel:  ret_from_fork_asm+0x1a/0x30
2024-06-16T16:09:44-07:00 helium kernel:  </TASK>
2024-06-16T16:09:44-07:00 helium kernel: Modules linked in: tls mptcp_diag xsk_diag tcp_diag udp_diag raw_diag inet_diag unix_diag af_packet_diag netlink_diag rpcrdma rdma_cm iw_cm ib_cm ib_core cls_u32 sch_htb ifb wireguard curve25519_x86_64 libchacha20poly1305 chacha_x86_64 poly1305_x86_64 libcurve25519_generic libchacha ip6_udp_tunnel udp_tunnel bridge stp llc cfg80211 rfkill ip6table_nat ip6table_mangle ip6t_REJECT nf_reject_ipv6 xt_nat xt_REDIRECT xt_MASQUERADE iptable_nat xt_connmark xt_TCPMSS iptable_mangle xt_multiport xt_helper ipt_REJECT nf_reject_ipv4 xt_NFLOG nfnetlink_log xt_limit xt_conntrack xt_tcpudp nf_nat_irc nf_conntrack_irc nf_nat_sip nf_conntrack_sip nf_nat_ftp nf_conntrack_ftp nf_nat_pptp nf_conntrack_pptp nf_nat nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 lz4hc lz4hc_compress ip6table_filter ip6_tables iptable_filter joydev coretemp kvm_intel iTCO_wdt intel_pmc_bxt gpio_ich iTCO_vendor_support ppdev kvm zram sha512_ssse3 i2c_i801 sha1_ssse3 i2c_smbus acpi_cpufreq lpc_ich intel_agp intel_gtt parport_pc parport
2024-06-16T16:09:44-07:00 helium kernel:  hid_logitech_hidpp mac_hid zfs(POE) spl(OE) vhost_scsi target_core_mod vhost_net tun vhost vhost_iotlb tap usbip_host usbip_core pcspkr it87(OE) hwmon_vid frandom(OE) nfsd sg crypto_user loop auth_rpcgss nfs_acl lockd grace sunrpc nfnetlink ip_tables x_tables vfat fat hid_logitech_dj btrfs ata_generic hid_generic pata_acpi blake2b_generic xor raid6_pq sha256_ssse3 usbhid pata_jmicron ext4 crc16 mbcache jbd2 dm_cache dm_thin_pool dm_persistent_data dm_bio_prison libcrc32c crc32c_generic dm_snapshot dm_bufio dm_mod radeon drm_ttm_helper ttm video wmi i2c_algo_bit drm_suballoc_helper drm_display_helper cec lz4 lz4_compress netconsole r8169 realtek mdio_devres libphy r8168(OE)
2024-06-16T16:09:44-07:00 helium kernel: CR2: 0000000000000120
2024-06-16T16:09:44-07:00 helium kernel: ---[ end trace 0000000000000000 ]---
2024-06-16T16:09:44-07:00 helium kernel: RIP: 0010:dataset_kstats_rename+0x10/0x30 [zfs]
2024-06-16T16:09:44-07:00 helium kernel: Code: 1f 84 00 00 00 00 00 0f 1f 00 90 90 90 90 90 90 90 90 90 90 90 90 90 90 90 90 66 0f 1f 00 0f 1f 44 00 00 48 8b 87 98 03 00 00 <48> 8b 80 20 01 00 00 8b 90 08 01 00 00 48 8b b8 00 01 00 00 e9 67
2024-06-16T16:09:44-07:00 helium kernel: RSP: 0018:ffffa06e886a3d38 EFLAGS: 00010292
2024-06-16T16:09:44-07:00 helium kernel: RAX: 0000000000000000 RBX: ffff88edacc7f000 RCX: 00000001861b8801
2024-06-16T16:09:44-07:00 helium kernel: RDX: 00000001861b8601 RSI: ffff88ee4072b100 RDI: ffff88ee212a7998
2024-06-16T16:09:44-07:00 helium kernel: RBP: ffff88edefc05000 R08: 0000000000000000 R09: 0000000000000000
2024-06-16T16:09:44-07:00 helium kernel: R10: 0000000000000000 R11: 0000000000000000 R12: ffff88ee212a7d60
2024-06-16T16:09:44-07:00 helium kernel: R13: ffff88edacc7f004 R14: ffff88ed96915280 R15: ffff88ee212a7800
2024-06-16T16:09:44-07:00 helium kernel: FS:  0000000000000000(0000) GS:ffff88ee81c80000(0000) knlGS:0000000000000000
2024-06-16T16:09:44-07:00 helium kernel: CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
2024-06-16T16:09:44-07:00 helium kernel: CR2: 0000000000000120 CR3: 000000004f110000 CR4: 00000000000406f0
2024-06-16T16:09:44-07:00 helium kernel: note: z_zvol[982] exited with irqs disabled
```
</details>

### PR DESCRIPTION
**Pull request by jgottula**
### Motivation and Context

Fixes #16274 (a regression introduced by #15486). The issue affects stable releases beginning with 2.2.3.

### Description

Adds a simple NULL check to `dataset_kstats_rename` so that when the function is called for a snapshot (which do not have kstats allocated for them, at least at present), nothing is done.

This fix should have the additional bonus of "just doing the right thing" in the event that things ever change such that kstats _are_ allocated for snapshots: in that case, the NULL check would no longer bail out of the function, and it would instead update the `dataset_name` kstat value for the zvol snapshot in question.

### How Has This Been Tested?
As mentioned over in #16274, I applied the patch here to my own local build of zfs pretty early on, and it was effective at preventing the bug from occurring: I was then able to rename snapshotted zvols with `snapdev=visible` as much as I wanted, without any kthread fatalities or zvol-related stuff hanging indefinitely.

The problem is consistent and very easy to reproduce. I provided some slightly more detailed steps over in the issue thread, but it's really just as simple as:
1. Create pool
2. Create zvol
3. Ensure the zvol has `snapdev=visible`
4. Ensure the zvol has one or more snapshots
5. Do `zfs rename` on the zvol

@asomers helpfully verified that the bug affects FreeBSD too, and that this patch fixes things appropriately there as well.

I suppose that I could also test a build of zfs where I disable the snapshot ignore logic at the beginning of `dataset_kstats_create`, to double-check the assertion I made earlier that this patch would "just work", and cause kstats-for-snapshots-of-zvols to reflect renames properly, in the event that such kstats should ever exist. I haven't actually gone ahead and done this though. Let me know if that's something you'd like me to do.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] **(N/A)** I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] **(See comment below)** I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] **(GitHub CI only)** I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

I haven't run the test suite myself, but I've let GitHub Actions run it.

### CODE REVIEW

**Comment by jgottula** [comment]
It was suggested to me here https://github.com/openzfs/zfs/issues/16274#issuecomment-2174658745, that I add a case to the test suite; and so I began looking into doing that. The goal, as I understand it, would basically be to ensure that the "renaming a zvol which has snaps and snapdev=visible" set of circumstances is actually covered, so that the original bug would have been caught and so that future bugs of a similar nature would also be noticed.

But seeing as this was a "kernel doing crashy things" sort of problem, it wasn't entirely clear to me how I should proceed, as I noted here https://github.com/openzfs/zfs/issues/16274#issuecomment-2174936103. What precisely should constitute a pass or a failure in this circumstance? In the context of this particular issue, failure manifested as the `zfs rename` command working, and the zvol having been successfully renamed after that point as far as e.g. the zfs userspace utils were concerned; but `dmesg` reflecting a null pointer dereference, and the `z_zvol` kernel thread no longer existing, and udev workers—attempting to update the `/dev/zvol/...` symlinks—blocking indefinitely and eventually being killed off, and so forth. So how would I even check for a kernel screwup in a test case; how would I do that in a platform-neutral manner; what if there wasn't a way to check that didn't involve hanging indefinitely; etc? And does it even actually make sense for a test case to "fail" in such a way that it screws up the kernel and requires a reboot in order for subsequent operations to not potentially hang forever?

And then, assuming that the answer to all of that is presumably just "yeah, no, you don't try to intentionally check for this sort of kernel problem in the test suite", then what sort of test case (and I suppose that by that I mainly mean, what sort of pass/fail criteria) would still make sense to do for the "renaming a zvol which has snaps and snapdev=visible" circumstance? Presumably it's still valuable to have a test which covers that situation; because even if, say, in the event of the original bug here, such a test might not have "failed" in the ordinary sense, it would still have triggered the kernel bug and caused that to be noticed, one way or another.

So yeah, some guidance on what makes the most sense to do here would be appreciated.

**Review by robn — APPROVED** [review]
This is obviously correct, given the number of similar checks, and that we don't set `dk_stats` for whole classes of datasets, and that as a general pattern we allow for kstats to fail to be creation.

With regard to the test case, I agree that this sort of "programmer is holding it wrong" is rarely worth testing for in a functional test suite. There is value in test cases that that test the feature, or combination of features, but you came here to fix a crash, and you have, and you've been very diligent with working to understand the problem and ensure that this doesn't just paper over some deeper problem.

In other words: great method, great outcome. Thank you :)

**Review by amotin — APPROVED** [review]

**Comment by jgottula** [comment]
@robn Oops, I think I accidentally re-requested a review by you just now. Should probably do this on a desktop computer, where I can actually see tooltips for buttons before I press them... 😄

**Review by asomers — APPROVED** [review]

**Review by robn — APPROVED** [review]

**Review by allanjude — APPROVED** [review]

**Review by tonyhutter — APPROVED** [review]
